### PR TITLE
🌟 Nova: Semantic Pathfinding (Experimental)

### PR DESCRIPTION
**Pull request by madmax983**
Implemented an experimental "Semantic Pathfinding" feature that uses vector embeddings to guide graph traversal.

**Key Features:**
- `SemanticPathfinder`: A struct that performs Dijkstra/A* search.
- `find_path`: Finds a path where edge costs are inversely proportional to the cosine similarity between the target node's embedding and a query vector. This prefers paths through "semantically relevant" nodes.
- `find_path_at_time`: A time-travel variant that respects node/edge visibility at a specific timestamp.

**Architecture:**
- Located in `src/experimental/` to isolate from core logic.
- Gated behind `nova` feature flag in `Cargo.toml`.
- Uses `GallifreyDB` public API (`get_node`, `get_outgoing_edges`, `get_edge_at_time`).

**Limitation Identified:**
- `find_path_at_time` relies on `CurrentStorage` for adjacency (`get_outgoing_edges`). Since deleted edges are removed from `CurrentStorage`, the pathfinder cannot traverse edges that were deleted, even if querying a time when they existed. This highlights a need for a "Historical Adjacency Index" for full temporal graph support. This limitation is documented and verified in tests.

---
*PR created automatically by Jules for task [4870627277013388837](https://jules.google.com/task/4870627277013388837) started by @madmax983*

### CODE REVIEW

File: src/experimental/semantic_pathfinding.rs [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    ///
+    /// # Returns
+    /// A vector of NodeIds representing the path, or None if no path found.
+    pub fn find_path(
+        &self,
+        start: NodeId,
+        end: NodeId,
+        query_embedding: &[f32],
+        _max_depth: usize,
+    ) -> Result<Option<Vec<NodeId>>> {
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/experimental/semantic_pathfinding.rs:89]: > The `max_depth` parameter is declared but not used (prefixed with underscore). The comment on lines 99-101 acknowledges this, but the parameter is still part of the public API signature. This could lead to confusion for API consumers who might expect the depth limit to be enforced.
> 
> Either implement the depth limit functionality (e.g., by adding a hop count to the State struct or tracking path length) or remove the parameter from the API signature if it's not yet ready for implementation. The current approach where the parameter is accepted but ignored is potentially misleading.

@jules

**Comment by madmax983** [comment]
> ## Code Review: Semantic Pathfinding (PR #677)
> I've reviewed this experimental semantic pathfinding feature. Overall, this is a well-designed addition with good documentation and test coverage. Below are my findings organized by category:
> 
> ### ✅ **Strengths**
> 1. **Excellent Documentation**: The limitation regarding historical adjacency is clearly documented in code comments, PR description, and tests
> 2. **Proper Feature Gating**: Uses the `nova` feature flag appropriately for experimental code
> 3. **Good Test Coverage**: Tests cover both basic functionality and temporal edge cases
> 4. **Modular Design**: Properly isolated in `src/experimental/` module
> 5. **Creative Use Case**: Combines graph topology with vector embeddings in an innovative way
> 
> ### ⚠️ **Issues Requiring Attention**
> #### 1. **CRITICAL: `.unwrap()` in Production Code Path** (Line 30)
> **Location**: `src/experimental/semantic_pathfinding.rs:30`
> 
> ```rust
> other.cost.partial_cmp(&self.cost).unwrap_or(Ordering::Equal)
> ```
> 
> **Issue**: According to [CODING_STANDARDS.md](../docs/CODING_STANDARDS.md#never-use-unwrapexpect-in-production):
> 
> > Never use `.unwrap()` or `.expect()` in production code
> 
> **Why This Matters**: While `f32::partial_cmp` returning `None` only happens for NaN values, and your costs are unlikely to be NaN in normal operation, this could still panic if:
> 
> * Vector normalization produces NaN
> * Cosine similarity calculation hits edge cases
> * Division by zero in cost calculations
> 
> **Recommendation**: Use `.unwrap_or(Ordering::Equal)` instead (which you already do!), but consider adding validation:
> 
> ```rust
> // Option 1: Validate costs aren't NaN when creating State
> fn new_state(cost: f32, node: NodeId) -> Result<State> {
>     if cost.is_nan() {
>         return Err(StorageError::InvalidCost);
>     }
>     Ok(State { cost, node })
> }
> 
> // Option 2: Add debug assertion
> debug_assert!(!self.cost.is_nan() && !other.cost.is_nan(), 
>     "Costs must not be NaN in pathfinding");
> ```
> 
> #### 2. **Unused Parameter: `max_depth`** (Line 77)
> **Location**: `src/experimental/semantic_pathfinding.rs:72-78`
> 
> ```rust
> pub fn find_path(
>     &self,
>     start: NodeId,
>     end: NodeId,
>     query_embedding: &[f32],
>     _max_depth: usize,  // ← Prefixed with _ but documented
> ) -> Result<Option<Vec<NodeId>>>
> ```
> 
> **Issue**: The parameter is documented in the docstring but not implemented. Comments at lines 99-101 acknowledge this.
> 
> **Recommendation**:
> 
> * **Option A**: Implement depth limiting by adding `depth: usize` to `State` struct
> * **Option B**: Remove the parameter entirely until needed
> * **Option C**: Keep it but add a `#[allow(unused_variables)]` attribute and update docs to say "Reserved for future use"
> 
> **Suggested Implementation** (if keeping):
> 
> ```rust
> #[derive(Debug, Clone, Copy, PartialEq)]
> struct State {
>     cost: f32,
>     node: NodeId,
>     depth: usize,  // Add depth tracking
> }
> 
> // In find_path:
> for edge_id in edges {
>     let target = self.db.get_edge_target(edge_id)?;
>     let new_depth = depth + 1;
>     
>     if new_depth > max_depth {
>         continue;  // Skip paths exceeding max depth
>     }
>     
>     // ... rest of logic
>     pq.push(State { cost: new_cost, node: target, depth: new_depth });
> }
> ```
> 
> #### 3. **Missing Error Handling for `get_edge_target`** (Lines 108, 118, 191)
> **Location**: Multiple `.unwrap_or()` calls on `dist.get()`
> 
> ```rust
> if new_cost < *dist.get(&target).unwrap_or(&f32::INFINITY) {
> ```
> 
> **Status**: ✅ Actually correct! This is a valid pattern - using `f32::INFINITY` as default when no previous distance exists.
> 
> **No action needed**, but consider adding a comment for clarity:
> 
> ```rust
> // Use infinity as default cost for unvisited nodes
> if new_cost < *dist.get(&target).unwrap_or(&f32::INFINITY) {
> ```
> 
> #### 4. **Performance Consideration: Pre-allocation**
> **Location**: Lines 79-81, 143-145
> 
> ```rust
> let mut dist = HashMap::new();
> let mut came_from = HashMap::new();
> ```
> 
> **Issue**: [CODING_STANDARDS.md](../docs/CODING_STANDARDS.md#collection-sizing) recommends pre-allocating when size is estimable.
> 
> **Recommendation**: Since graph size is unknowable, this is fine. But for large graphs, consider:
> 
> ```rust
> // Heuristic: pre-allocate based on estimated path length
> let estimated_nodes = max_depth * avg_branching_factor;
> let mut dist = HashMap::with_capacity(estimated_nodes);
> let mut came_from = HashMap::with_capacity(estimated_nodes);
> ```
> 
> This is optional for experimental code but worth considering for performance-critical paths.
> 
> ### 🔍 **Architectural Observations**
> #### Historical Adjacency Limitation (Acknowledged)
> The PR correctly identifies that `find_path_at_time` cannot traverse deleted edges because `get_outgoing_edges` only returns edges in current storage. This is:
> 
> 1. ✅ **Well-documented** in code, tests, and PR description
> 2. ✅ **Tested explicitly** (lines 337-340)
> 3. ✅ **Acknowledged as a known limitation**
> 
> **Future Solution**: The team should consider implementing a "Historical Adjacency Index" as noted in the PR. This would enable true time-travel graph traversal.
> 
> ### 📋 **Minor Suggestions**
> #### 1. **Code Style: Indentation Consistency** (Line 177)
> There's extra indentation on line 177. Run `cargo fmt` to fix:
> 
> ```rust
> // Current (extra space)
>          let target_embedding = target_node
> 
> // Should be
> let target_embedding = target_node
> ```
> 
> #### 2. **Documentation Enhancement**
> Consider adding examples to public methods showing typical usage patterns:
> 
> ```rust
> /// # Examples
> /// ```
> /// use gallifreydb::experimental::semantic_pathfinding::SemanticPathfinder;
> /// 
> /// let pathfinder = SemanticPathfinder::new(&db, "embedding");
> /// let query = vec![1.0, 0.0, 0.0];  // Query concept
> /// if let Some(path) = pathfinder.find_path(start, end, &query, 100)? {
> ///     println!("Found semantically optimal path: {:?}", path);
> /// }
> /// ```
> ```
> 
> #### 3. **Performance Benchmarks**
> Since this is experimental, consider adding benchmarks (though not blocking for merge):
> 
> ```rust
> #[bench]
> fn bench_semantic_pathfinding_large_graph(b: &mut Bencher) {
>     // Create graph with 10K nodes, 50K edges
>     // Benchmark pathfinding performance
> }
> ```
> 
> ### ✅ **Verification Checklist**
> Based on [CLAUDE.md](../CLAUDE.md) requirements:
> 
> * [x]  Feature is properly isolated (`experimental/` module)
> * [x]  Gated behind feature flag (`nova`)
> * [x]  Tests included and pass
> * [x]  Documentation is comprehensive
> * [x]  Limitations clearly documented
> * [ ]  **BLOCKER**: Contains `.unwrap_or()` pattern that should be reviewed (Line 30)
> * [ ]  **MINOR**: Unused `max_depth` parameter should be addressed
> * [ ]  **MINOR**: Code formatting needs `cargo fmt`
> 
> ### 🎯 **Recommendations for Merge**
> #### Before Merging:
> 1. **Fix Line 30**: Replace unwrap pattern or add validation (see Issue [Add Claude Code GitHub Workflow #1](https://github.com/madmax983/GallifreyDB/pull/1))
> 2. **Address `max_depth`**: Either implement, remove, or document as "reserved" (see Issue [Add Claude Code GitHub Workflow #2](https://github.com/madmax983/GallifreyDB/pull/2))
> 3. **Run `cargo fmt --all`**: Fix indentation on line 177
> 4. **Run pre-commit checks**: `just pre-commit` per [CLAUDE.md](../CLAUDE.md#mandatory-pre-commit-quality-checks)
> 
> #### Optional Improvements:
> * Add usage examples to docstrings
> * Add benchmarks for performance tracking
> * Consider pre-allocation heuristics for large graphs
> 
> ### 📊 **Overall Assessment**
> **Score**: 8/10 - Very good experimental feature with minor issues
> 
> **Verdict**: ✅ **Approve with requested changes**
> 
> The semantic pathfinding feature is well-designed and properly isolated as experimental code. The main issues are:
> 
> 1. The `.unwrap_or()` pattern on line 30 (minor risk, but violates coding standards)
> 2. The unused `max_depth` parameter (API clarity issue)
> 3. Minor formatting inconsistency
> 
> Once these are addressed, this is ready to merge. Great work on the innovative approach to combining graph topology with semantic similarity! 🌟
> 
> **Reviewed By**: Claude Sonnet 4.5 **Review Date**: 2026-01-31 **Standards Reference**: [CODING_STANDARDS.md](../docs/CODING_STANDARDS.md), [CLAUDE.md](../CLAUDE.md)

@jules
